README.md

# [MOVED] Analytics JS integration woopra

**This repository has been moved to the open source [Analytics JS Integrations](https://github.com/segmentio/analytics.js-integrations).**

If you want to fork, contribute, or open issues, please do it in the new repository. Existing issues/pull requests will be addressed in the new location.

* [New location](https://github.com/segmentio/analytics.js-integrations/tree/master/integrations/woopra)
* Last release for this repository: [2.0.1](https://github.com/segment-integrations/analytics.js-integration-woopra/releases/tag/2.0.1)
* [Commit](https://github.com/segmentio/analytics.js-integrations/commit/a724dd909f7d2da8a76d4efb94a81efe00a2feaf)
* [Previous version of this README](README-OLD.md)

Released under the [MIT license](LICENSE).


HISTORY.md
2.0.0 / 2016-06-21
==================

  * Remove Duo compatibility
  * Add CI setup (coverage, linting, cross-browser compatibility, etc.)
  * Update eslint configuration

1.0.6 / 2016-05-07
==================

  * Bump Analytics.js core, tester, integration to use Facade 2.x

1.0.5 / 2015-11-10
==================

  * Update a.js woopra integration to handle traits with dates

1.0.4 / 2015-11-10
==================

  * use components to remain browser compliant

1.0.3 / 2015-09-10
==================

  * stringify nested obj in track.properties

1.0.2 / 2015-06-30
==================

  * Replace analytics.js dependency with analytics.js-core

1.0.1 / 2015-06-24
==================

  * Bump analytics.js-integration version

1.0.0 / 2015-06-09
==================

  * Initial commit :sparkles:


test/index.test.js

var Analytics = require('analytics.js-core').constructor;
var integration = require('analytics.js-integration');
var tester = require('analytics.js-integration-tester');
var sandbox = require('clear-env');
var Woopra = require('../lib/');

describe('Woopra', function() {
  var analytics;
  var woopra;
  var options = {
    domain: 'x',
    outgoingTracking: false
  };

  beforeEach(function() {
    analytics = new Analytics();
    woopra = new Woopra(options);
    analytics.use(Woopra);
    analytics.use(tester);
    analytics.add(woopra);
  });

  afterEach(function() {
    analytics.restore();
    analytics.reset();
    woopra.reset();
    sandbox();
  });

  it('should have the right settings', function() {
    analytics.compare(Woopra, integration('Woopra')
      .global('woopra')
      .option('domain', '')
      .option('cookieName', 'wooTracker')
      .option('cookieDomain', null)
      .option('cookiePath', '/')
      .option('ping', true)
      .option('pingInterval', 12000)
      .option('idleTimeout', 300000)
      .option('downloadTracking', true)
      .option('outgoingTracking', true)
      .option('outgoingIgnoreSubdomain', true)
      .option('downloadPause', 200)
      .option('outgoingPause', 400)
      .option('ignoreQueryUrl', true)
      .option('hideCampaign', false));
  });

  describe('before loading', function() {
    beforeEach(function() {
      analytics.stub(woopra, 'load');
    });

    describe('#initialize', function() {
      it('should create a woopra object', function() {
        analytics.assert(!window.woopra);
        analytics.initialize();
        analytics.page();
        analytics.assert(window.woopra);
      });

      it('should configure woopra', function() {
        analytics.initialize();
        analytics.page();
        analytics.deepEqual(window.woopra._e, [
          ['config', 'domain', 'x'],
          ['config', 'outgoing_tracking', false],
          ['config', 'cookie_name', 'wooTracker'],
          ['config', 'cookie_path', '/'],
          ['config', 'ping', true],
          ['config', 'ping_interval', 12000],
          ['config', 'idle_timeout', 300000],
          ['config', 'download_tracking', true],
          ['config', 'outgoing_ignore_subdomain', true],
          ['config', 'download_pause', 200],
          ['config', 'outgoing_pause', 400],
          ['config', 'ignore_query_url', true],
          ['config', 'hide_campaign', false]
        ]);
      });

      it('should not send options if they are null, or empty', function() {
        woopra.options.domain = '';
        woopra.options.cookieName = '';
        woopra.options.cookiePath = null;
        woopra.options.ping = null;
        woopra.options.pingInterval = null;
        woopra.options.idleTimeout = null;
        woopra.options.downloadTracking = null;
        woopra.options.outgoingTracking = null;
        woopra.options.outgoingIgnoreSubdomain = null;
        woopra.options.downloadPause = '';
        woopra.options.outgoingPause = '';
        woopra.options.ignoreQueryUrl = null;
        woopra.options.hideCampaign = null;
        analytics.initialize();
        analytics.page();
        analytics.deepEqual([], window.woopra._e);
      });

      it('should call #load', function() {
        analytics.initialize();
        analytics.page();
        analytics.called(woopra.load);
      });
    });
  });

  describe('loading', function() {
    it('should load', function(done) {
      analytics.load(woopra, done);
    });
  });

  describe('after loading', function() {
    beforeEach(function(done) {
      analytics.once('ready', done);
      analytics.initialize();
      analytics.page();
    });

    describe('#page', function() {
      beforeEach(function() {
        analytics.stub(window.woopra, 'track');
      });

      it('should send a page view', function() {
        analytics.page();
        analytics.called(window.woopra.track, 'pv', {
          path: window.location.pathname,
          referrer: document.referrer,
          title: document.title,
          search: window.location.search,
          url: window.location.protocol + '//' + window.location.hostname + (window.location.port ? ':' + window.location.port : '') + window.location.pathname
        });
      });

      it('should send a title', function() {
        analytics.page({ title: 'title' });
        analytics.called(window.woopra.track, 'pv', {
          title: 'title',
          path: window.location.pathname,
          referrer: document.referrer,
          search: window.location.search,
          url: window.location.protocol + '//' + window.location.hostname + (window.location.port ? ':' + window.location.port : '') + window.location.pathname
        });
      });

      it('should prefer a name', function() {
        analytics.page('name', { title: 'title' });
        analytics.called(window.woopra.track, 'pv', {
          title: 'name',
          name: 'name',
          path: window.location.pathname,
          referrer: document.referrer,
          search: window.location.search,
          url: window.location.protocol + '//' + window.location.hostname + (window.location.port ? ':' + window.location.port : '') + window.location.pathname
        });
      });

      it('should prefer a category and name', function() {
        analytics.page('category', 'name', { title: 'title' });
        analytics.called(window.woopra.track, 'pv', {
          title: 'category name',
          category: 'category',
          name: 'name',
          path: window.location.pathname,
          referrer: document.referrer,
          search: window.location.search,
          url: window.location.protocol + '//' + window.location.hostname + (window.location.port ? ':' + window.location.port : '') + window.location.pathname
        });
      });
    });

    describe('#identify', function() {
      beforeEach(function() {
        analytics.stub(window.woopra, 'identify');
      });

      it('should send an id', function() {
        analytics.identify('id');
        analytics.called(window.woopra.identify, { id: 'id' });
      });

      it('should send traits', function() {
        analytics.identify({ trait: true });
        analytics.called(window.woopra.identify, { trait: true });
      });

      it('should send an id and traits', function() {
        analytics.identify('id', { trait: true });
        analytics.called(window.woopra.identify, { id: 'id', trait: true });
      });

      it('should alias the name properly', function() {
        analytics.identify('id', {
          firstName: 'firstName',
          lastName: 'lastName'
        });
        analytics.called(window.woopra.identify, {
          name: 'firstName lastName',
          firstName: 'firstName',
          lastName: 'lastName',
          id: 'id'
        });
      });

      it('should should convert trait dates to unix timestamp in milliseconds', function() {
        analytics.identify('id', { testdate: '2015-11-04T09:20:22Z' });
        analytics.called(window.woopra.identify, { id: 'id', testdate: 1446628822000 });
      });
    });

    describe('#track', function() {
      beforeEach(function() {
        analytics.stub(window.woopra, 'track');
      });

      it('should send an event', function() {
        analytics.track('event');
        analytics.called(window.woopra.track, 'event');
      });

      it('should send properties', function() {
        analytics.track('event', { property: 'Property' });
        analytics.called(window.woopra.track, 'event', { property: 'Property' });
      });

      it('should stringify nested objects', function(){
        analytics.track('event', {
          products: [
              {
                sku: '45790-32',
                name: 'Monopoly: 3rd Edition'
              },
              {
                sku: '46493-32',
                name: 'Uno Card Game'
              }
            ],
          orderId: 1
        });
        analytics.called(window.woopra.track, 'event', { products: '[{"sku":"45790-32","name":"Monopoly: 3rd Edition"},{"sku":"46493-32","name":"Uno Card Game"}]', orderId: 1 });
      });
    });
  });
});


lib/index.js
'use strict';

/**
 * Module dependencies.
 */

var each = require('component-each');
var integration = require('@segment/analytics.js-integration');
var toSnakeCase = require('to-snake-case');
var is = require('is-type');
var foldl = require('@ndhoule/foldl');
var json = require('json3');
var isostring = require('isostring');
var time = require('unix-time');

/**
 * Expose `Woopra` integration.
 */

var Woopra = module.exports = integration('Woopra')
  .global('woopra')
  .option('domain', '')
  .option('cookieName', 'wooTracker')
  .option('cookieDomain', null)
  .option('cookiePath', '/')
  .option('ping', true)
  .option('pingInterval', 12000)
  .option('idleTimeout', 300000)
  .option('downloadTracking', true)
  .option('outgoingTracking', true)
  .option('outgoingIgnoreSubdomain', true)
  .option('downloadPause', 200)
  .option('outgoingPause', 400)
  .option('ignoreQueryUrl', true)
  .option('hideCampaign', false)
  .tag('<script src="//static.woopra.com/js/w.js">');

/**
 * Initialize.
 *
 * http://www.woopra.com/docs/setup/javascript-tracking/
 */

Woopra.prototype.initialize = function() {
  /* eslint-disable */
  (function(){var i, s, z, w = window, d = document, a = arguments, q = 'script', f = ['config', 'track', 'identify', 'visit', 'push', 'call'], c = function(){var i, self = this; self._e = []; for (i = 0; i < f.length; i++){(function(f){self[f] = function(){self._e.push([f].concat(Array.prototype.slice.call(arguments, 0))); return self; }; })(f[i]); } }; w._w = w._w || {}; for (i = 0; i < a.length; i++){ w._w[a[i]] = w[a[i]] = w[a[i]] || new c(); } })('woopra');
  /* eslint-enable */

  this.load(this.ready);
  each(this.options, function(key, value) {
    key = toSnakeCase(key);
    if (value == null) return;
    if (value === '') return;
    window.woopra.config(key, value);
  });
};

/**
 * Loaded?
 *
 * @return {Boolean}
 */

Woopra.prototype.loaded = function() {
  return !!(window.woopra && window.woopra.loaded);
};

/**
 * Page.
 *
 * @param {String} category (optional)
 */

Woopra.prototype.page = function(page) {
  var props = page.properties();
  var name = page.fullName();
  if (name) props.title = name;
  window.woopra.track('pv', props);
};

/**
 * Identify.
 *
 * @param {Identify} identify
 */

Woopra.prototype.identify = function(identify) {
  var traits = identify.traits();

  // Woopra likes timestamps in milliseconds
  // Ref: https://www.woopra.com/docs/manual/configure-schema/
  each(traits, function(key, val) {
    if (isostring(val) || is.date(val)) {
      traits[key] = time(val) * 1000;
    }
  });

  if (identify.name()) traits.name = identify.name();
  // `push` sends it off async
  window.woopra.identify(traits).push();
};

/**
 * Track.
 *
 * @param {Track} track
 */

Woopra.prototype.track = function(track) {
  window.woopra.track(track.event(), stringifyNested(track.properties()));
};

/**
 * Stringify nested objects.
 *
 * Undocumented aspect of Woopra's API, but apparently required. Breaks
 * on `Completed Order` `properties.products`.
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */

function stringifyNested(obj) {
  return foldl(function(results, value, key) {
    if (is.array(obj[key])) {
      results[key] = json.stringify(obj[key]);
    } else {
      results[key] = obj[key]; 
    }
    return results;
  }, {}, obj);
}
